File: examples/sigaction.c

#include <stdio.h>
#include <signal.h>
#include <time.h>
#include <unistd.h>
#include <string.h>
#include <sys/signal.h>

static void signal_handler(int sig_num) {
  printf("signal handle %d\n", sig_num);
}

static const int kSigLen = 14;

int main() {
  const int kSignals[] = {
      SIGHUP,
      SIGINT,
      SIGQUIT,
      SIGILL,
      SIGABRT,
      SIGBUS,
      SIGFPE,
      SIGUSR1,
      SIGUSR2,
      SIGPIPE,
      SIGALRM,
      SIGTERM,
      SIGCHLD,
      SIGPOLL,
  };

  struct sigaction sa;
  bzero(&sa, sizeof(struct sigaction));
  sa.sa_handler = signal_handler;
  sa.sa_flags = SA_RESTART;

  for (int i = 0; i < kSigLen; ++i) {
      int sig_num = kSignals[i];
      int ret = sigaction(sig_num, &sa, NULL);
      printf("register signal handler for %d\n", sig_num);
    }

    const int pid = getpid();
    printf("pid: %d\n", pid);

    while (1) {
      struct timespec ts = { .tv_sec = 10, .tv_nsec = 0};
      int ret = nanosleep(&ts, NULL);
    }

  return 0;
}

File: examples/sigaction.rs
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by Apache-2.0 License that can be found
// in the LICENSE file.

fn signal_handler(sig_num: i32) {
    println!("signal handler {sig_num}");
}

const SIGNALS: [i32; 14] = [
    nc::SIGSEGV,
    nc::SIGHUP,
    nc::SIGINT,
    nc::SIGQUIT,
    nc::SIGILL,
    nc::SIGABRT,
    nc::SIGBUS,
    nc::SIGFPE,
    nc::SIGUSR1,
    nc::SIGUSR2,
    nc::SIGPIPE,
    nc::SIGALRM,
    nc::SIGTERM,
    nc::SIGCHLD,
];

fn main() {
    let sa = nc::sigaction_t {
        sa_handler: signal_handler as nc::sighandler_t,
        sa_flags: nc::SA_RESTART | nc::SA_RESTORER,
        sa_restorer: nc::restore::get_sa_restorer(),
        ..Default::default()
    };
    for sig_num in SIGNALS {
        let ret = unsafe { nc::rt_sigaction(sig_num, Some(&sa), None) };
        assert!(ret.is_ok());
        println!("register signal handler for {sig_num}");
    }

    let pid = unsafe { nc::getpid() };
    println!("pid: {}", pid);

    loop {
        let t = nc::timespec_t {
            tv_sec: 10,
            tv_nsec: 0,
        };
        unsafe {
            let _ret = nc::nanosleep(&t, None);
        }
    }
}
